### PR TITLE
GCP: allow stop/autostop for spot VMs.

### LINKED ISSUE #2837: Consider allowing `sky stop` on GCP spot clusters
Not allowing `sky stop` on spot clusters was a decision made a long time ago, and it was modeled after AWS behavior. AWS would throw an error like this
```
You can't stop the Spot Instance 'i-xxx' because it is associated with a one-time Spot Instance request. You can only stop Spot Instances associated with persistent Spot Instance requests.
```
On GCP, however, they do [allow](https://cloud.google.com/compute/docs/instances/preemptible#preemption-process) the disks to be preserved:
```
You can access and recover data from any persistent disks that are attached to the instance, but those disks still incur storage charges until you delete them.
```

A user raised the above and ranked it as "nice to have".

### PR DESCRIPTION
**Pull request by concretevitamin**
Fixes #2837.

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_stop_gcp_spot` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/core.py [hunk 1]
```diff
@@ -327,9 +328,11 @@ def stop(cluster_name: str, purge: bool = False) -> None:
         cloud = handle.launched_resources.cloud
         cloud.check_features_are_supported(
             {clouds.CloudImplementationFeatures.STOP})
-        if handle.launched_resources.use_spot:
+        # Allow GCP spot to be stopped since it preserves disk:
+        #   https://cloud.google.com/compute/docs/instances/preemptible#preemption-process  # pylint: disable=line-too-long
+        if handle.launched_resources.use_spot and not cloud.is_same_cloud(
+                clouds.GCP()):
```
Comment (Michaelvll) [inline sky/core.py:334]: How about we make this as a `CloudImplementationFeatures.STOP_SPOT_INSTANCE`, so we don't have to do a cloud specific check here?

File: sky/core.py [hunk 2]
```diff
@@ -437,26 +440,29 @@ def autostop(
...
+            f'{operation} cluster {cluster_name!r} with TPU VM Pod '
+            'is not supported.')
+
     if not isinstance(backend, backends.CloudVmRayBackend):
         raise exceptions.NotSupportedError(
             f'{operation} cluster {cluster_name!r} with backend '
             f'{backend.__class__.__name__!r} is not supported.')
-    elif handle.launched_resources.use_spot and not down and not is_cancel:
+    elif (handle.launched_resources.use_spot and not down and not is_cancel and
+          not handle.launched_resources.cloud.is_same_cloud(clouds.GCP())):
```
Comment (Michaelvll) [inline sky/core.py:456]: Similarly, should we make this one of the `CloudImplementationFeatures`?

**Review by Michaelvll — COMMENTED** [review]
Thanks for adding the support of stopping a spot instance on GCP @concretevitamin! The code looks mostly good to me! Left two nits.

File: sky/core.py (hunk 2 above)
Comment (concretevitamin) [reply sky/core.py]: Good call, done.

File: sky/core.py (hunk 1 above)
Comment (concretevitamin) [reply sky/core.py]: Good call, done.

File: sky/core.py [hunk 3]
```diff
@@ -437,26 +447,39 @@ def autostop(
...
-    if tpu_utils.is_tpu_vm_pod(handle.launched_resources):
-        # Reference:
-        # https://cloud.google.com/tpu/docs/managing-tpus-tpu-vm#stopping_a_with_gcloud  # pylint: disable=line-too-long
-        raise exceptions.NotSupportedError(
-            f'{operation} cluster {cluster_name!r} with TPU VM Pod '
-            'is not supported.')
-
     # Check autostop is implemented for cloud
     cloud = handle.launched_resources.cloud
     if not down and idle_minutes >= 0:
```
Comment (Michaelvll) [inline sky/core.py:485]: code gardening: Seems `CloudImplementationFeatures.AUTOSTOP` does not provide additional info than `CloudImplmentationFeatures.STOP`, maybe we should consider having them combined.

nit: Also, what do you think to have this line combined with L462-467?

**Review by Michaelvll — APPROVED** [review]
Thanks for the fix @concretevitamin! It looks mostly good to me with a comment for a refactoring of the `check_features_are_supported` function.

File: sky/core.py (hunk 3 above)
Comment (concretevitamin) [reply sky/core.py:487]: Good call, removed `CloudImplementationFeatures.AUTOSTOP`.

> nit: Also, what do you think to have this line combined with L462-467?

Hmm, I find it harder to reason since the call can raise. Separating STOP_SPOT_INSTANCE and STOP seems clearer?

**Review by concretevitamin — COMMENTED** [review]
Made some changes. PTAL.

File: sky/backends/cloud_vm_ray_backend.py [hunk 4]
```diff
@@ -4278,6 +4278,19 @@ def set_autostop(self,
...
+            assert (handle.launched_resources is not None and
+                    handle.launched_resources.cloud is not None), handle
+            if handle.launched_resources.use_spot:
+                # This can be triggered by, for example:
+                #   sky launch --cloud aws --use-spot --cpus 2+ -i0 -y
+                # The cluster will be UP, the launch exited with code 1, and
+                # any stage after PRE_EXEC is not executed.
+                cloud = handle.launched_resources.cloud
+                cloud.check_features_are_supported(
+                    {clouds.CloudImplementationFeatures.STOP_SPOT_INSTANCE})
```
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:4292]: Does this mean that we will only raise the error after the cluster is launched? This can be missed as people may just run `sky launch` and leave it there, it might be better to raise the error earlier before the launching actually starts.

What we did for other partially supported features is that we check the features during the failover and skip the clouds that do not support the feautures requested. Can we add this `STOP_SPOT_INSTANCE` into the requested features:
https://github.com/skypilot-org/skypilot/blob/1bb59fd25b7b2de9ccb65c52c42a356145aaddb7/sky/backends/cloud_vm_ray_backend.py#L2207-L2208

This is to fix the following case:
1. `sky launch --use-spot -i 0 echo hi` where the failover first go over AWS and then go to GCP.
2. In current implementation, we do not check the supported feature before we launch the cluster causing the error being raised after the cluster is UP, which may leave the cluster on forever, until the users check `sky status -r` themselves.
In such case, it might be better to just get rid of AWS from the failover list and directly try to run the job on GCP.

**Review by Michaelvll — COMMENTED** [review]
Thank you for the fix @concretevitamin! The code looks mostly good to me, except for when we raise an error for the STOP_SPOT_INSTANCE being unsupported.

File: sky/backends/cloud_vm_ray_backend.py (hunk 4 above)
Comment (concretevitamin) [reply sky/backends/cloud_vm_ray_backend.py:4292]: We can do that, with a bigger refactoring in execute.py#execute(): instead of 

https://github.com/skypilot-org/skypilot/blob/1bb59fd25b7b2de9ccb65c52c42a356145aaddb7/sky/execution.py#L306-L308

We can change `requested_features` into a dict of `Resources -> set(features)`. This is because for a DAG node with `any_of`, e.g.,
```
[Task(run=<empty>)
  resources: {AWS([Spot], cpus=2+), GCP(cpus=2+)}]
```
The first resource will request STOP_SPOT while the second doesn't. 

One problem is this doesn't feel clean -- this refactoring is only for STOP_SPOT_INSTANCE while the rest of CloudImplementationFeatures are all handled cleanly. But I do not see a better way.

Wdyt?

Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:4292]: Hmm, when we do `sky launch -i` for a DAG, we probably need all the resources in the DAG to support the stopping.

Probably, the refactoring you mentioned before by using `cloud.check_feature_supported(resources, features)` is needed, i.e., we remove the `STOP_SPOT_INSTANCE` feature, but only decide whether a launchable resources supports `STOP` based on the field the resources have.

**Comment by Michaelvll** [comment]
@concretevitamin I refactored the handling for checking supported features. Please take a look : )

File: sky/core.py [hunk 5]
```diff
@@ -317,26 +317,22 @@ def stop(cluster_name: str, purge: bool = False) -> None:
...
+                handle.launched_resources,
+                {clouds.CloudImplementationFeatures.STOP})
+        except exceptions.NotSupportedError as e:
             raise exceptions.NotSupportedError(
                 f'{colorama.Fore.YELLOW}Stopping cluster '
                 f'{cluster_name!r}... skipped.{colorama.Style.RESET_ALL}\n'
-                '  Stopping spot instances is not supported as the attached '
-                'disks will be lost.\n'
+                '  Stopping instances is not supported for '
+                f'{handle.launched_resources}.\n'
```
Comment (concretevitamin) [inline sky/core.py]: This message may be a bit hard to parse. How about we special case to test `handle.launched_resources.use_spot` and print the original "Stopping spot instances is currently not supported on <cloud>". and for other resources cases keep the current message?

File: sky/core.py [hunk 6]
```diff
@@ -437,31 +433,28 @@ def autostop(
...
-            {clouds.CloudImplementationFeatures.AUTOSTOP})
+        try:
+            cloud.check_features_are_supported(
+                handle.launched_resources,
+                {clouds.CloudImplementationFeatures.STOP})
+        except exceptions.NotSupportedError as e:
+            raise exceptions.NotSupportedError(
+                f'{colorama.Fore.YELLOW}Scheduling autostop on cluster '
+                f'{cluster_name!r}...skipped.{colorama.Style.RESET_ALL}\n'
+                '  Stopping instances is not supported for '
```
Comment (concretevitamin) [inline sky/core.py]: likewise

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll for the refactoring! Failover works correctly now.

I pushed some minor changes in messages. PTAL.

File: sky/clouds/cloud.py [hunk 7]
```diff
@@ -457,7 +443,8 @@ def need_cleanup_after_preemption(
 
     @classmethod
     def check_features_are_supported(
-            cls, requested_features: Set[CloudImplementationFeatures]) -> None:
+            cls, resources: 'resources_lib.Resources',
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/clouds/cloud.py:446]: Updated. Thanks!

File: sky/clouds/gcp.py [hunk 8]
```diff
@@ -809,8 +817,6 @@ def need_cleanup_after_preemption(self,
         # you must delete it and create a new one ..."
         # See: https://cloud.google.com/tpu/docs/preemptible#tpu-vm
 
-        # pylint: disable=import-outside-toplevel
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/clouds/gcp.py]: I suppose it was for avoiding circular import, but since we have updated a lot of places, it seems fine to import it at the top level

**Review by Michaelvll — APPROVED** [review]
LGTM.  :)
